### DISCUSSION TITLE
Module Design Methodology Question

### DISCUSSION THREAD

**Discussion by DrtyByt** [oqtane/oqtane.framework#2057] [General]
My question can be summarized as, should complex modules be built using a single Oqtane module with a menu to select the type of interface into the module or should the module be done as separate pieces and then forgo the DB constructs that are auto scaffolded with the module creation?
In depth explanation with some pros and cons:
Using a single module with a custom interface selector would allow this module to be used in the multi-tenant setup and even allow multiple instances of the software in a tenant. This would get more complex because you need to do a few things to get this to function correctly. 1) You have to create a custom selector type to determine what interface into the software you want that module to represent. 2) Additional permission options would be needed so that you can allow the types of ui that would be allowed by a user claim (most likely via role). 3) Those permissions need to be manually wired up in the ui. This approach seems like the best way to model the application into Oqtane, but I wanted to see if there was something else coming in the works or if someone else has a different Idea.
Using multiple modules that all connect to the same Data Repository no matter the module ID.
Using this method means that you would create a separate module for each interface (UI Element) being used for the application. This makes for much smaller to ingest modules instead of one monolithic one. The issue comes into all of these modules needing to be interconnected to the same set of data. Without a ton of custom code this would not be multi-tenant capable. This would flood the modules list as well. This is much simpler on the permissions front as you don't need to create any custom permissions or ingest them.

**Comment by markdav-is** [comment]
You can create a "mega-module" that share a repo and an API and other things as one solution.  In the UI layer you can create multiple sub-folders and each one can have it's own module definition.  this can be deployed as a single package and you will see all the modules defined therein in the dropdown when adding modules to pages.  This method might work for your situation.
